src/GafferScene/FilterQuery.cpp

```diff
@@ -0,0 +1,346 @@
+			}
+		}
+	}
+	else if( output == exactMatchPlug() )
+	{
+		ComputeNode::hash( output, context, h );
+		matchPlug()->hash( h );
+	}
+	else if( output == descendantMatchPlug() )
+	{
+		ComputeNode::hash( output, context, h );
+		matchPlug()->hash( h );
+	}
+	else if( output == ancestorMatchPlug() )
+	{
+		ComputeNode::hash( output, context, h );
+		matchPlug()->hash( h );
+	}
+	else if( output == closestAncestorPlug() )
+	{
+		const string location = locationPlug()->getValue();
+		if( !location.empty() )
+		{
+			ScenePlug::PathScope scope( context, ScenePlug::stringToPath( location ) );
+			h = closestAncestorInternalPlug()->hash();
+		}
+		else
+		{
+			h = output->defaultHash();
+		}
+	}
+	else if( output == closestAncestorInternalPlug() )
+	{
+		if( scenePlug()->existsPlug()->getValue() )
+		{
+			const int m = filterPlug()->match( scenePlug() );
+			if( m & PathMatcher::ExactMatch )
+			{
+				ComputeNode::hash( output, context, h );
+				const ScenePlug::ScenePath path = context->get<ScenePlug::ScenePath>( ScenePlug::scenePathContextName );
+				h.append( path.data(), path.size() );
+				return;
+			}
+			else if( m & PathMatcher::AncestorMatch )
+			{
+				ScenePlug::ScenePath path = context->get<ScenePlug::ScenePath>( ScenePlug::scenePathContextName );
+				if( path.size() )
+				{
+					path.pop_back();
+					ScenePlug::PathScope scope( context, path );
+					h = closestAncestorInternalPlug()->hash();
+					return;
+				}
+			}
+		}
+		h = output->defaultHash();
+	}
+}
+
+void FilterQuery::compute( Gaffer::ValuePlug *output, const Gaffer::Context *context ) const
+{
+	if( output == matchPlug() )
+	{
+		unsigned match = IECore::PathMatcher::NoMatch;
+		const string location = locationPlug()->getValue();
+		if( !location.empty() )
+		{
+			ScenePlug::PathScope scope( context, ScenePlug::stringToPath( location ) );
+			if( scenePlug()->existsPlug()->getValue() )
+			{
+				match = filterPlug()->match( scenePlug() );
+			}
+		}
+		static_cast<IntPlug *>( output )->setValue( match );
+	}
+	else if( output == exactMatchPlug() )
+	{
+		static_cast<BoolPlug *>( output )->setValue(
+			matchPlug()->getValue() & IECore::PathMatcher::ExactMatch
+		);
+	}
+	else if( output == descendantMatchPlug() )
+	{
+		static_cast<BoolPlug *>( output )->setValue(
+			matchPlug()->getValue() & IECore::PathMatcher::DescendantMatch
+		);
+	}
+	else if( output == ancestorMatchPlug() )
+	{
+		static_cast<BoolPlug *>( output )->setValue(
+			matchPlug()->getValue() & IECore::PathMatcher::AncestorMatch
+		);
+	}
+	else if( output == closestAncestorPlug() )
+	{
+		const string location = locationPlug()->getValue();
+		if( !location.empty() )
+		{
+			ScenePlug::PathScope scope( context, ScenePlug::stringToPath( location ) );
+			output->setFrom( closestAncestorInternalPlug() );
+		}
+		else
+		{
+			output->setToDefault();
+		}
+	}
+	else if( output == closestAncestorInternalPlug() )
+	{
+		string result;
+		if( scenePlug()->existsPlug()->getValue() )
```

Comment:
If someone uses a FilterQuery deep in the hierarchy at a location that isn't filtered, then existsPlug() will be evaluated at many parent locations, which doesn't seem necessary?  Probably isn't important, but there seems to be a small argument for moving the existsPlug check into closestAncestorPlug?  Might be more important in hash()?

Reply:
Makes sense - aeab136.